<a href="https://colab.research.google.com/github/Sandeeppasala/EVA_S5/blob/master/EVA4_Session_F1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms

In [0]:
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, padding=1) #input -? OUtput? RF
        self.conv2 = nn.Conv2d(32, 64, 3, padding=1)
        self.pool1 = nn.MaxPool2d(2, 2)
        self.conv3 = nn.Conv2d(64, 128, 3, padding=1)
        self.conv4 = nn.Conv2d(128, 256, 3, padding=1)
        self.pool2 = nn.MaxPool2d(2, 2)
        self.conv5 = nn.Conv2d(256, 512, 3)
        self.conv6 = nn.Conv2d(512, 1024, 3)
        self.conv7 = nn.Conv2d(1024, 10, 3)

    def forward(self, x):
        x = self.pool1(F.relu(self.conv2(F.relu(self.conv1(x)))))
        x = self.pool2(F.relu(self.conv4(F.relu(self.conv3(x)))))
        x = F.relu(self.conv6(F.relu(self.conv5(x))))
        x = (self.conv7(x))
        x = x.view(-1, 10)
        return F.log_softmax(x)

In [3]:
!pip install torchsummary
from torchsummary import summary
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")
model = Net().to(device)
summary(model, input_size=(1, 28, 28))

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1           [-1, 32, 28, 28]             320
            Conv2d-2           [-1, 64, 28, 28]          18,496
         MaxPool2d-3           [-1, 64, 14, 14]               0
            Conv2d-4          [-1, 128, 14, 14]          73,856
            Conv2d-5          [-1, 256, 14, 14]         295,168
         MaxPool2d-6            [-1, 256, 7, 7]               0
            Conv2d-7            [-1, 512, 5, 5]       1,180,160
            Conv2d-8           [-1, 1024, 3, 3]       4,719,616
            Conv2d-9             [-1, 10, 1, 1]          92,170
Total params: 6,379,786
Trainable params: 6,379,786
Non-trainable params: 0
----------------------------------------------------------------
Input size (MB): 0.00
Forward/backward pass size (MB): 1.51
Params size (MB): 24.34
Estimated Total Size (MB): 25.85
-------------------------------------

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:20: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.


In [0]:


torch.manual_seed(1)
batch_size = 128

kwargs = {'num_workers': 1, 'pin_memory': True} if use_cuda else {}
train_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=True, download=True,
                    transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))
                    ])),
    batch_size=batch_size, shuffle=True, **kwargs)
test_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=False, transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))
                    ])),
    batch_size=batch_size, shuffle=True, **kwargs)


In [0]:
from tqdm import tqdm
def train(model, device, train_loader, optimizer, epoch):
    model.train()
    pbar = tqdm(train_loader)
    for batch_idx, (data, target) in enumerate(pbar):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        pbar.set_description(desc= f'loss={loss.item()} batch_id={batch_idx}')


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)
    print("\n EPOCH::", epoch)
    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))

In [6]:

model = Net().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)

for epoch in range(1, 15):
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)

  0%|          | 0/469 [00:00<?, ?it/s]


 EPOCH:: 1

Test set: Average loss: 0.0640, Accuracy: 9801/10000 (98.01%)



  0%|          | 0/469 [00:00<?, ?it/s]


 EPOCH:: 2

Test set: Average loss: 0.0460, Accuracy: 9846/10000 (98.46%)



  0%|          | 0/469 [00:00<?, ?it/s]


 EPOCH:: 3

Test set: Average loss: 0.0306, Accuracy: 9903/10000 (99.03%)



  0%|          | 0/469 [00:00<?, ?it/s]


 EPOCH:: 4

Test set: Average loss: 0.0251, Accuracy: 9914/10000 (99.14%)



  0%|          | 0/469 [00:00<?, ?it/s]


 EPOCH:: 5

Test set: Average loss: 0.0247, Accuracy: 9918/10000 (99.18%)



  0%|          | 0/469 [00:00<?, ?it/s]


 EPOCH:: 6

Test set: Average loss: 0.0288, Accuracy: 9905/10000 (99.05%)



  0%|          | 0/469 [00:00<?, ?it/s]


 EPOCH:: 7

Test set: Average loss: 0.0260, Accuracy: 9922/10000 (99.22%)



  0%|          | 0/469 [00:00<?, ?it/s]


 EPOCH:: 8

Test set: Average loss: 0.0314, Accuracy: 9904/10000 (99.04%)



  0%|          | 0/469 [00:00<?, ?it/s]


 EPOCH:: 9

Test set: Average loss: 0.0320, Accuracy: 9906/10000 (99.06%)



  0%|          | 0/469 [00:00<?, ?it/s]


 EPOCH:: 10

Test set: Average loss: 0.0303, Accuracy: 9919/10000 (99.19%)



  0%|          | 0/469 [00:00<?, ?it/s]


 EPOCH:: 11

Test set: Average loss: 0.0345, Accuracy: 9907/10000 (99.07%)



  0%|          | 0/469 [00:00<?, ?it/s]


 EPOCH:: 12

Test set: Average loss: 0.0281, Accuracy: 9918/10000 (99.18%)



  0%|          | 0/469 [00:00<?, ?it/s]


 EPOCH:: 13

Test set: Average loss: 0.0363, Accuracy: 9910/10000 (99.10%)



loss=4.121661186218262e-05 batch_id=468: 100%|██████████| 469/469 [00:21<00:00, 21.90it/s]



 EPOCH:: 14

Test set: Average loss: 0.0318, Accuracy: 9918/10000 (99.18%)



Target:

Get the set-up right

Set Transforms

Set Data Loader

Set Basic Working Code

Set Basic Training  & Test Loop

Results:

Parameters: 6.3M

Best Training Accuracy: 100%

Best Test Accuracy: 99.2%

Analysis:

Extremely Heavy Model 

Model is over-fitting

Model can be further improved by reducing the number of parameters.


